File: apps/mobile/plugins/autofill-service/android/CryptoBridge.kt
package com.keykeykey.app

import com.goterl.lazysodium.LazySodiumAndroid
import com.goterl.lazysodium.SodiumAndroid
import com.goterl.lazysodium.interfaces.AEAD
import com.goterl.lazysodium.interfaces.PwHash

/**
 * Argon2id parameters matching @keykeykey/core Argon2Params.
 */
data class Argon2Params(
    val t: Int,
    val m: Int,
    val p: Int,
    val dkLen: Int,
)

/**
 * Cryptographic constants matching @keykeykey/core/crypto/constants.
 */
object CryptoConstants {
    /** 256-bit symmetric key length (DEK and KEK). */
    const val KEY_SIZE = 32
    /** 192-bit nonce for XChaCha20-Poly1305. */
    const val NONCE_SIZE = 24
    /** 128-bit Poly1305 authentication tag. */
    const val TAG_SIZE = 16
}

/**
 * Native crypto bridge using lazysodium for XChaCha20-Poly1305 and Argon2id.
 *
 * Mirrors the envelope encryption model from @keykeykey/core:
 * - decrypt: XChaCha20-Poly1305 with managed nonce (nonce prepended to ciphertext)
 * - deriveKEK: Argon2id key derivation
 * - unwrapDEK: decrypt wrapped DEK and validate length
 */
object CryptoBridge {

    private val lazySodium = LazySodiumAndroid(SodiumAndroid())

    /**
     * Decrypt ciphertext produced by @noble/ciphers managedNonce XChaCha20-Poly1305.
     *
     * Input format: [24B nonce][ciphertext][16B tag]
     * libsodium expects: nonce separate, ciphertext with tag appended.
     *
     * @param ciphertext The full ciphertext with prepended nonce
     * @param key 32-byte symmetric key
     * @return Decrypted plaintext
     * @throws IllegalArgumentException if key size is wrong or ciphertext too short
     * @throws SecurityException if decryption/authentication fails
     */
    fun decrypt(ciphertext: ByteArray, key: ByteArray): ByteArray {
        require(key.size == CryptoConstants.KEY_SIZE) {
            "Key must be ${CryptoConstants.KEY_SIZE} bytes, got ${key.size}"
        }
        require(ciphertext.size > CryptoConstants.NONCE_SIZE + CryptoConstants.TAG_SIZE) {
            "Ciphertext too short: ${ciphertext.size} bytes"
        }

        val nonce = ciphertext.copyOfRange(0, CryptoConstants.NONCE_SIZE)
        val encryptedWithTag = ciphertext.copyOfRange(CryptoConstants.NONCE_SIZE, ciphertext.size)

        val plaintext = ByteArray(encryptedWithTag.size - CryptoConstants.TAG_SIZE)
        val plaintextLen = longArrayOf(0)

        val success = lazySodium.cryptoAeadXChaCha20Poly1305IetfDecrypt(
            plaintext,
            plaintextLen,
            null, // nsec (unused)
            encryptedWithTag,
            encryptedWithTag.size.toLong(),
            null, // ad
            0L, // adLen
            nonce,
            key,
        )

        if (!success) {
            throw SecurityException("Decryption failed: authentication tag mismatch")
        }

        return plaintext.copyOf(plaintextLen[0].toInt())
    }

    /**
     * Derive a KEK from a password and salt using Argon2id.
     *
     * @param password The master password
     * @param salt 16-byte salt
     * @param params Argon2id tuning parameters
     * @return Derived key of params.dkLen bytes
     * @throws IllegalArgumentException if p != 1 (libsodium limitation)
     */
    fun deriveKEK(password: String, salt: ByteArray, params: Argon2Params): ByteArray {
        require(params.p == 1) {
            "libsodium Argon2id only supports p=1, got p=${params.p}"
        }

        val passwordBytes = password.toByteArray(Charsets.UTF_8)
        val output = ByteArray(params.dkLen)

        try {
            val result = lazySodium.cryptoPwHash(
                output,
                output.size.toLong(),
                passwordBytes,
                passwordBytes.size,
                salt,
                params.t.toLong(),
                params.m.toLong() * 1024, // m is in KiB, libsodium wants bytes
                PwHash.Alg.PWHASH_ALG_ARGON2ID13.value.toInt(),
            )

            if (!result) {
                throw SecurityException("Argon2id key derivation failed")
            }

            return output
        } finally {
            passwordBytes.fill(0)
        }
    }

    /**
     * Unwrap (decrypt) a wrapped DEK using a KEK and validate the result is 32 bytes.
     *
     * @param wrappedDEK The encrypted DEK (nonce + ciphertext + tag)
     * @param kek 32-byte Key Encryption Key
     * @return The 32-byte DEK
     * @throws SecurityException if decryption fails or DEK is not 32 bytes
     */
    fun unwrapDEK(wrappedDEK: ByteArray, kek: ByteArray): ByteArray {
        val dek = decrypt(wrappedDEK, kek)
        if (dek.size != CryptoConstants.KEY_SIZE) {
            dek.fill(0)
            throw SecurityException(
                "Unwrapped DEK is ${dek.size} bytes, expected ${CryptoConstants.KEY_SIZE}"
            )
        }
        return dek
    }
}

File: apps/mobile/targets/credential-provider/CryptoBridge.swift
import Foundation
import Sodium
import Clibsodium

// MARK: - Errors

/// Errors that can occur during cryptographic operations in the credential provider.
enum CryptoError: Error, LocalizedError {
    case decryptionFailed
    case invalidCiphertext(String)
    case invalidKeySize(Int)
    case unsupportedParallelism(String)
    case argon2Failed

    var errorDescription: String? {
        switch self {
        case .decryptionFailed:
            return "Decryption failed: authentication tag verification failed or wrong key"
        case .invalidCiphertext(let reason):
            return "Invalid ciphertext: \(reason)"
        case .invalidKeySize(let size):
            return "Invalid key size: expected \(CryptoConstants.keySize) bytes, got \(size)"
        case .unsupportedParallelism(let reason):
            return "Unsupported parallelism: \(reason)"
        case .argon2Failed:
            return "Argon2id key derivation failed"
        }
    }
}

// MARK: - Constants

/// Cryptographic constants matching the TypeScript @noble/ciphers format.
///
/// Wire format: `[24-byte nonce][ciphertext][16-byte Poly1305 tag]`
struct CryptoConstants {
    static let keySize = 32
    static let nonceSize = 24
    static let tagSize = 16
    static let saltSize = 16
    /// Total overhead added by managed-nonce XChaCha20-Poly1305: nonce + tag.
    static let overhead = nonceSize + tagSize // 40
}

// MARK: - Argon2 Parameters

/// Argon2id tuning parameters matching the TypeScript `Argon2Params` interface.
struct Argon2Params {
    /// Time cost (number of iterations / opslimit).
    let t: Int
    /// Memory cost in KiB (memlimit = m * 1024).
    let m: Int
    /// Parallelism (must be 1 for libsodium's `crypto_pwhash`).
    let p: Int
    /// Derived key length in bytes.
    let dkLen: Int
}

// MARK: - CryptoBridge

/// Thin wrapper around libsodium that matches the TypeScript `@noble/ciphers` binary format exactly.
/// Read/decrypt only -- this bridge is used by the credential provider extension to unlock vaults.
struct CryptoBridge {
    private let sodium = Sodium()

    /// Decrypt ciphertext produced by the TypeScript core's `encrypt()` function.
    ///
    /// Expected input format: `[24-byte nonce][ciphertext][16-byte Poly1305 tag]`
    ///
    /// libsodium's `xchacha20poly1305ietf.decrypt` expects `[ciphertext][tag]` as the
    /// authenticated ciphertext, so we strip the leading 24-byte nonce and pass the rest.
    ///
    /// - Parameters:
    ///   - ciphertext: The full wire-format ciphertext (nonce + ciphertext + tag).
    ///   - key: A 32-byte symmetric key.
    /// - Returns: The decrypted plaintext.
    /// - Throws: `CryptoError.invalidKeySize` if key is not 32 bytes.
    /// - Throws: `CryptoError.invalidCiphertext` if data is too short.
    /// - Throws: `CryptoError.decryptionFailed` if authentication fails.
    func decrypt(_ ciphertext: Data, key: Data) throws -> Data {
        guard key.count == CryptoConstants.keySize else {
            throw CryptoError.invalidKeySize(key.count)
        }
        guard ciphertext.count > CryptoConstants.overhead else {
            throw CryptoError.invalidCiphertext(
                "too short: expected > \(CryptoConstants.overhead) bytes, got \(ciphertext.count)"
            )
        }

        let nonce = ciphertext.subdata(in: 0 ..< CryptoConstants.nonceSize)
        // Remainder is [ciphertext][tag] — exactly what libsodium expects
        let authenticatedCiphertext = ciphertext.subdata(
            in: CryptoConstants.nonceSize ..< ciphertext.count
        )

        let nonceBytes = Array(nonce)
        let ciphertextBytes = Array(authenticatedCiphertext)
        let keyBytes = Array(key)

        guard let plaintext = sodium.aead.xchacha20poly1305ietf.decrypt(
            authenticatedCipherText: ciphertextBytes,
            secretKey: keyBytes,
            nonce: nonceBytes
        ) else {
            throw CryptoError.decryptionFailed
        }

        return Data(plaintext)
    }

    /// Unwrap (decrypt) a wrapped DEK using the KEK and validate the result is 32 bytes.
    ///
    /// - Parameters:
    ///   - wrappedDEK: The encrypted DEK (output of TypeScript `wrapDEK`).
    ///   - kek: The 32-byte Key Encryption Key derived from the master password.
    /// - Returns: The 32-byte Data Encryption Key.
    /// - Throws: `CryptoError` if decryption or validation fails.
    func unwrapDEK(_ wrappedDEK: Data, kek: Data) throws -> Data {
        let dek = try decrypt(wrappedDEK, key: kek)
        guard dek.count == CryptoConstants.keySize else {
            throw CryptoError.invalidCiphertext(
                "unwrapped DEK must be \(CryptoConstants.keySize) bytes, got \(dek.count)"
            )
        }
        return dek
    }

    /// Derive a KEK from a master password using Argon2id via libsodium's `crypto_pwhash`.
    ///
    /// - Parameters:
    ///   - password: The user's master password (UTF-8 string).
    ///   - salt: A 16-byte salt from the vault header.
    ///   - params: Argon2id tuning parameters (t, m, p, dkLen).
    /// - Returns: The derived key (KEK) of length `params.dkLen`.
    /// - Throws: `CryptoError.unsupportedParallelism` if `params.p != 1`.
    /// - Throws: `CryptoError.argon2Failed` if derivation fails.
    func deriveKEK(password: String, salt: Data, params: Argon2Params) throws -> Data {
        guard params.p == 1 else {
            throw CryptoError.unsupportedParallelism(
                "libsodium crypto_pwhash only supports p=1, got p=\(params.p)"
            )
        }

        var passwordBytes = Array(password.utf8)
        defer {
            for i in passwordBytes.indices {
                passwordBytes[i] = 0
            }
        }

        let saltBytes = Array(salt)
        var output = [UInt8](repeating: 0, count: params.dkLen)

        // crypto_pwhash maps: t -> opslimit, m * 1024 -> memlimit (KiB to bytes)
        let result = crypto_pwhash(
            &output,
            UInt64(params.dkLen),
            passwordBytes.map { Int8(bitPattern: $0) },
            UInt64(passwordBytes.count),
            saltBytes,
            UInt64(params.t),
            Int(params.m) * 1024,
            Int32(crypto_pwhash_ALG_ARGON2ID13)
        )

        guard result == 0 else {
            throw CryptoError.argon2Failed
        }

        return Data(output)
    }
}
